src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs

```diff
@@ -218,6 +236,57 @@ internal void GenerateMoveNext(BoundStatement body, MethodSymbol moveNextMethod)
             F.CloseMethod(newBody);
         }
 
+        private BoundBlock GenerateExceptionHandling(LocalSymbol exceptionLocal)
+        {
+            // this.state = finishedState
+            BoundExpressionStatement assignFinishedState =
+                F.Assignment(F.Field(F.This(), stateField), F.Literal(StateMachineStates.FinishedStateMachine));
+
+            if (_asyncIteratorInfo == null)
+            {
+                return F.Block(
+                    // this.state = finishedState
+                    assignFinishedState,
+                    // builder.SetException(ex)
+                    F.ExpressionStatement(
+                        F.Call(
+                            F.Field(F.This(), _asyncMethodBuilderField),
+                            _asyncMethodBuilderMemberCollection.SetException,
+                            F.Local(exceptionLocal))),
+                    GenerateReturn(false));
+            }
+            else
+            {
+                // this.state = finishedState
+                // if (promiseIsActive)
```

Comment:
Would this be better as an exception filter?

Reply:
I didn't understand the suggestion. 
The `this.state = finishedState;` portion always needs to be run, so if we use an exception filter, I think we'd need a second `catch` clause to save the state.

---
In reply to: [210742184](https://github.com/dotnet/roslyn/pull/28218#discussion_r210742184) [](ancestors = 210742184)